### DISCUSSION TITLE
Update Spritesheet Texture in 5.x

### DISCUSSION THREAD

**Discussion by Benholio** [pixijs/pixijs#7716] [Q&A]
In 4.x, we have implemented a lazy-loaded system for spritesheets that involves first creating a placeholder image and then swapping in the real image once it has been loaded. This was accomplished by calling ```spriteSheet.baseTexture.loadSource(img)```.

In 5.x, you can no longer load the source of a baseTexture more than once. It was discussed in an [issue](https://github.com/pixijs/pixijs/issues/5609#issuecomment-489096129) that it might be implemented, but it looks like it never was.

My question - does anyone have advice for how to work around this? I've tried:

1. Hacking around the baseTexture change to force a new source to be loaded
2. Creating a new Resource and assigning it to the baseTexture
3. Create a new baseTexture and assigning it to the SpriteSheet

None of these seem to be supported or easy to hack around in 5.x. I haven't dug into the library code to see how hard it would be to change baseTexture to allow reloading the source, but I assume there are major issues with that or it would have been done already. Please correct me if I'm wrong.

Thanks!

**Comment by ivanpopelyshev** [comment]
2. with a hack , because there's `throw` somewhere about two resources

**Reply by ivanpopelyshev** [reply]
>  I haven't dug into the library code to see how hard it would be to change baseTexture to allow reloading the source, but I assume there are major issues with that or it would have been done already

there are no major issues with that. We made texture resources because we cant possibly imagine all the ways people load those textures. Yes, even then sometimes we assume that "people wont need to change resource per baseTexture", that particular throw wasnt my idea :)

**Reply by ivanpopelyshev** [reply]
maybe "correct" approach is to make your own ImageResource that first loads temporary tex to videomem , and only when its ready, does the real upload. One resource, just logic differs from pixi.

**Reply by Benholio** [reply]
Thanks Ivan!

**Reply by Benholio** [reply]
I've had some time to work on these suggestions, thanks!

I tried the "#2" solution first, to remove the throw about reassigning resources, but I ran into issues. It does seem like other parts of the library are assuming that the resource on a baseTexture does not change.

Now I'm trying the "correct" approach to create my own resource, and I have a couple of questions about that approach.

1. Can you give any examples of how my custom resource would upload texture data to the GPU? In the existing resource classes, the upload() method is called internally from elsewhere inside of pixi. I'm not sure how to get the correct parameters to call from outside (or from within the new resource)
2. In my current usage, I am creating textures with Texture.fromLoader, which always creates an ImageResource. For testing, I just hacked that method to create my custom resource (ImageLazyResource), but is there a more proper way I should be doing that? I'm trying to avoid modifying the pixijs source as much as possible to make future upgrades less painful.

Thanks for your time!

**Reply by ivanpopelyshev** [reply]
> is called internally from elsewhere inside of pixi.
from TextureSystem

when you setSize() , texture becomes valid, and at its first bind() TextureSystem will call upload.

> I am creating textures with Texture.fromLoader
pixi doesnt have a way to swap defeault ImageResource with something else, so yeah, you have to hack that thing.
